### DISCUSSION TITLE
How to use ASCII characters to search accented characters in consult-line?

### DISCUSSION THREAD

**Discussion by rodrigo-morales-1** [minad/consult#1057] [Q&A] [answered]
Let me explain a minimal working example:

I saved the contents of the code block below to `/tmp/a.el`.

```elisp
(use-package consult
  :bind
  ("C-l" . consult-line))
```

I saved the contents of the code block below to `/tmp/a.txt`. Note that the last four lines have the accented characters: `á` (U+00E1, LATIN SMALL LETTER A WITH ACUTE) and `é` (U+00E9, LATIN SMALL LETTER E WITH ACUTE). These two characters are part of the Spanish alphabet.

```text
a1
a2
e1
e2
á1
á2
é1
é2
```

I started Emacs using the following command:

```sh
$ emacs --quick --eval='(load "/tmp/a.el")' --eval='(find-file "/tmp/a.txt")'
```

I pressed `M-x consult-line RET`. I inserted `a` in the minibuffer and I pressed `TAB`. The buffer `*Completions*` was shown on top of the minibuffer, its content is shown below

```text
Click on a completion to select it.
In this buffer, type RET to select the completion near point.

2 possible completions:
1 a1
2 a2
```

I want the results to be:

```text
Click on a completion to select it.
In this buffer, type RET to select the completion near point.

2 possible completions:
1 a1
2 a2
5 á1
6 á2
```

That is, I want to use ASCII characters to search accented characters. How to accomplish this behavior with `consult-line`?

<a id="orgb192a0d"></a>

# Motivation

I'm currently editing files in a language that use the accented characters (e.g. áéíóú). I wish I could use ASCII characters to match accented characters when using `consult-line`.

# Additional information

I found some sexps on the Internet that might be relevant to this issue.

The following sexp was retrieved from [this comment](https://emacs.stackexchange.com/questions/71937/find-notes-with-accented-characters-when-searching-for-the-base-character-like#comment116292_71937) at Emacs Stack Exchange.

```elisp
(char-fold-to-regexp "a")
```

    \(?:a[̀-̄̆-̨̣̥̊̌̏̑]\|[aªà-åāăąǎǟǡǻȁȃȧᵃḁạảấầẩẫậắằẳẵặₐⓐａ𝐚𝑎𝒂𝒶𝓪𝔞𝕒𝖆𝖺𝗮𝘢𝙖𝚊]\)

The following sexp was retrieved from [this answer](https://emacs.stackexchange.com/a/75531) at Emacs Stack Exchange.

```elisp
(char-to-string (car (get-char-code-property ?á 'decomposition)))
```

    a

**Comment by minad** [comment, marked as answer]
Hint: Search for `char-fold-to-regexp` in the Orderless documentation.

**Reply by rodrigo-morales-1** [reply]
Thanks, after searching related information on Orderless. I found two solutions:

# Solution 1: Use `%` as a suffix for words in the prompt of `consult-line`

I saved the contents of the following code block to `/tmp/a.el`.

```elisp
(use-package vertico
  :config
  (vertico-mode 1))

(use-package orderless
  :init
  (setq completion-styles '(orderless)))

(use-package consult
  :bind
  ("C-l" . consult-line))
```

I saved the contents of the following code block to `/tmp/a.txt`.

```text
día mano fruta
avión batería vegetal
mercado día tienda
```

I started Emacs using the following command:

```sh
emacs --quick --eval='(load "/tmp/a.el")' --eval='(find-file "/tmp/a.txt")'
```

I searched `dia%` and the lines candidates were: `1 día mano fruta` and `3 mercado día tienda`.


<a id="org90b01fd"></a>

# Solution 2: Add `char-fold-to-regexp` to `orderless-matching-styles`

I learned this from [this comment](https://www.reddit.com/r/emacs/comments/11lyg9k/comment/jbglvtl/) at Reddit.

I saved the contents of the following code block to `/tmp/a.el`.

```elisp
(use-package vertico
  :config
  (vertico-mode 1))

(use-package orderless
  :init
  (setq completion-styles '(orderless))
  (setq orderless-matching-styles
        (list #'orderless-literal #'orderless-regexp #'char-fold-to-regexp)))

(use-package consult
  :bind
  ("C-l" . consult-line))
```

I saved the contents of the following code block to `/tmp/a.txt`.

```text
día mano fruta
avión batería vegetal
mercado día tienda
```

I started Emacs using the following command:

```sh
emacs --quick --eval='(load "/tmp/a.el")' --eval='(find-file "/tmp/a.txt")'
```

I searched `dia` and the candidates were: `1 día mano fruta` and `3 mercado día tienda`.

**Reply by minad** [reply]
Note that solution 2 is slow due the costly regexps, therefore we don't use this configuration by default in Orderless.
